starter-code/functions-and-arrays.js

```diff
@@ -130,8 +183,127 @@ var howManyDog = howManyTimes(words, "dog");
+
+/*function greatestProduct (matrix) {
+  var max_product = 0
+  for (var row = 0; row < matrix.length; row++) {
+    for (var column = 0; column < matrix[row].length; column++) {
+      //current element => matrix[row][column]
+
+      // =========================
+      // Special case: corners
+      // =========================
+
+      // corner 1
+      if (row === 0 && column == 0) {
+        var product = matrix[row + 1][column] * matrix[row][column + 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // corner 2
+      if (row === 0 && column == matrix[row].length - 1) {
+        var product = matrix[row + 1][column] * matrix[row][column - 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // corner 3
+      if (row === matrix.length - 1 && column == 0) {
+        var product = matrix[row - 1][column] * matrix[row][column + 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // corner 4
+      if (row === matrix.length - 1 && column == matrix[row].length - 1) {
+        var product = matrix[row - 1][column] * matrix[row][column - 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // =========================
+      // Special case: Rows
+      // =========================
+      if (row === 0) {
+        var product = matrix[row][column - 1] * matrix[row + 1][column] * matrix[row][column + 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      if (row === matrix.length - 1) {
+        var product = matrix[row][column - 1] * matrix[row - 1][column] * matrix[row][column + 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // =========================
+      // Special case: Columns
+      // =========================
+      if (column === 0 && row != 0 && row != matrix.length - 1) {
+        var product = matrix[row - 1][column] * matrix[row][column + 1] * matrix[row + 1][column]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      if (column === matrix[row].length - 1 && row != 0 && row != matrix.length - 1) {
+        var product = matrix[row - 1][column] * matrix[row][column + 1] * matrix[row + 1][column]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+
+      // =========================
+      // General case
+      // =========================
+      if (column != 0 && row != 0 && column != matrix[row].length - 1 && row != matrix.length - 1) {
+        var product = matrix[row + 1][column] * matrix[row - 1][column] * matrix[row][column - 1] * matrix[row][column + 1]
+        if (product > max_product) {
+          max_product = product
+        }
+      }
+  }}
+  return max_product
+}*/
+
 function greatestProduct (matrix) {
```

Comment:
This is the most elegant solution you could have ended up with. Very nice work!
Notice however that if a cell holds a zero, the associated condition inside the `if`s will be false. To be perfectly correct, you can check for `undefined` by doing for example `if(matrix[row - 1] === undefined)`instead of `if(matrix[row - 1])`.